### PR TITLE
Create plot_finale_2.py

### PR DESCRIPTION
**Pull request by lpizzino**
plot percentage pure data vs raw data

### CODE REVIEW

**Comment by fgrosa** [comment]
Ciao Lorenzo, 
thanks a lot for the script! I have a copule of comments, could you implement them and push them before we merge the PR?

1) I suggest to change the name of the file to something that is more evocative, such as "plot_purity.py".

2) the script works only with MC (since in data we do not have the MC truth), so I suggest to remove the possibility to run on data files.

3) could you please add the log scale in the x-axis? ( you can have a look at plot_hist2d.py )

4) I suggest to enlarge the figure with the purities and put either 4x2 or 3x3 panels because as it is now the figures are difficult to understand.

Thanks!

**Comment by lpizzino** [comment]
Ciao Fabrizio,

appena posso faccio le modifiche richieste.

Volevo chiederti se mettendo i dati sulle x in scala logaritmica, dovessi cambiare anche l errore sulle x.

Buona giornata

Lorenzo

Ottieni Outlook per Android<https://aka.ms/ghei36>

________________________________

**Comment by fgrosa** [comment]
Ciao Lorenzo, 

there is no need to change the uncertainty on the x-axis when using a log scale.

Thanks!
Ciao,
Fabrizio

File: plot_purity.py [hunk 1]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
...
-        plt.subplot(4,2,z+1)
+        plt.subplot(2,2,z+1)
         dfraw = (pd.DataFrame(ptraw[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['sum']}))
         dfpure = (pd.DataFrame(ptpure[j], columns=['p']).assign(Bin=lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['sum']}))
         dfrawcount = (pd.DataFrame(ptraw[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['count']}))
         dfpurecount = (pd.DataFrame(ptpure[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['count']}))
         yer = np.divide(np.sqrt(np.asarray(dfpurecount)*(1 - np.divide(dfpurecount,dfrawcount))),dfrawcount)
-        plt.scatter(width3, np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
-        plt.errorbar(width3, np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
+        plt.scatter(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
```
Comment (fgrosa) [inline plot_purity.py]: remove

File: plot_purity.py [hunk 2]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
     z=0
     for j in labels:
         plt.figure(3)
-        plt.subplot(4,2,z+1)
```
Comment (fgrosa) [inline plot_purity.py]: before for loop
```py
fig1, axes1 = plt.subplot(2,2)
```

File: plot_purity.py [hunk 3]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
...
         dfraw = (pd.DataFrame(ptraw[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['sum']}))
         dfpure = (pd.DataFrame(ptpure[j], columns=['p']).assign(Bin=lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['sum']}))
         dfrawcount = (pd.DataFrame(ptraw[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['count']}))
         dfpurecount = (pd.DataFrame(ptpure[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['count']}))
         yer = np.divide(np.sqrt(np.asarray(dfpurecount)*(1 - np.divide(dfpurecount,dfrawcount))),dfrawcount)
-        plt.scatter(width3, np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
-        plt.errorbar(width3, np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
+        plt.scatter(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
+        plt.errorbar(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)
```
Comment (fgrosa) [inline plot_purity.py]:
```py
axes[index[0], index[1]].errorbar(width3, np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5, marker='o')
```

File: plot_purity.py [hunk 4]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
...
         dfpurecount = (pd.DataFrame(ptpure[j], columns=['p']).assign(Bin = lambda x: pd.cut(x.p, bins=width2)).groupby(['Bin']).agg({'p': ['count']}))
         yer = np.divide(np.sqrt(np.asarray(dfpurecount)*(1 - np.divide(dfpurecount,dfrawcount))),dfrawcount)
-        plt.scatter(width3, np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
-        plt.errorbar(width3, np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
+        plt.scatter(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
+        plt.errorbar(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
         plt.title(nomi[z])
-        plt.axis([0,13,0,1.3])    
-        plt.xlabel('P')
+        plt.axis([-1,7,0,1.1])
```
Comment (fgrosa) [inline plot_purity.py]: remove

File: plot_purity.py [hunk 5]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
...
         yer = np.divide(np.sqrt(np.asarray(dfpurecount)*(1 - np.divide(dfpurecount,dfrawcount))),dfrawcount)
-        plt.scatter(width3, np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
-        plt.errorbar(width3, np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
+        plt.scatter(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), marker='o', c=color[z])
+        plt.errorbar(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
         plt.title(nomi[z])
-        plt.axis([0,13,0,1.3])    
-        plt.xlabel('P')
+        plt.axis([-1,7,0,1.1])    
+        plt.xlabel('log(P)')
```
Comment (fgrosa) [inline plot_purity.py]:
```py
axes[index[0], index[1]].set_xlabel('p (GeV/c)')
axes[index[0], index[1]].set_ylabel('purity')
axes[index[0], index[1]].set_xscale('log')
```

File: plot_purity.py [hunk 6]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
     z=0
     for j in labels:
```
Comment (fgrosa) [inline plot_purity.py]:
```py
for index in product([0,1], [0,1]):
```

File: plot_purity.py [hunk 7]
```diff
@@ -61,20 +61,42 @@ def plotfinale2(file):
...
+        plt.errorbar(np.log(width3), np.asarray(np.divide(dfpure,dfraw)), xerr = xer , yerr = np.asarray(yer), fmt='.k', capsize=3, elinewidth=0.5)    
         plt.title(nomi[z])
-        plt.axis([0,13,0,1.3])    
-        plt.xlabel('P')
+        plt.axis([-1,7,0,1.1])    
+        plt.xlabel('log(P)')
         plt.ylabel('% pure')
         plt.tight_layout()
         z = z+1
+        if z==4:
```
Comment (fgrosa) [inline plot_purity.py]: same as upper for loop

File: plot_purity.py [hunk 8]
```diff
@@ -152,16 +174,10 @@ def plotfinale2(file):
 PARSER = argparse.ArgumentParser(description='Arguments to pass')
 PARSER.add_argument('dir', metavar='text', default='.',
                     help='flag to search in directory, remember put all paths!')
-PARSER.add_argument('--data', action='store_true',
-                    help='flag to search and visualize data file')
 PARSER.add_argument('--mc', action='store_true',
```
Comment (fgrosa) [inline plot_purity.py]: it can be removed since there can only work with MC

File: plot_purity.py [hunk 9]
```diff
@@ -152,16 +174,10 @@ def plotfinale2(file):
...
                     help='flag to search and visualize Monte Carlo file')
 ARGS = PARSER.parse_args()
 
-if not ARGS.data and not ARGS.mc:
-    print('You need to specificy if mc or data files')
-    exit()
-
 if ARGS.mc:
```
Comment (fgrosa) [inline plot_purity.py]: remove if

File: plot_purity.py [hunk 10]
```diff
@@ -49,7 +49,7 @@ def plotfinale2(file):
     labels = {'e': 'electrons', 'pi': 'pions', 'kaons': 'kaons', 'p': 'protons',
               'He3': 'He', 'triton': 'triton', 'deuterons': 'deuterons'}
```
Comment (fgrosa) [inline plot_purity.py]: add on top of the file
```py
from itertools import product
```

**Pull request merged**